File: android/app/src/test/java/app/prompter/tablet/protocol/AndroidWireCompatibilityTest.kt
package app.prompter.tablet.protocol

import org.junit.Assert.assertEquals
import org.junit.Test

class AndroidWireCompatibilityTest {
    private val codec = ProtocolCodec { 1_700_000_000_000 }
    private fun fixture(name: String) = requireNotNull(javaClass.classLoader?.getResource("protocol/$name")).readText().trim()
    private fun <T> encode(type: String, payload: T, serializer: kotlinx.serialization.KSerializer<T>) =
        codec.encodeEnvelope(type, payload, serializer, "server-wire", "android-pending", "android-wire", 0)

    @Test fun `real Android codec emits Windows-compatible pairing bytes`() {
        val base = PairRequest(requestId = "request-wire", deviceId = "device-wire", deviceName = "Galaxy Tab A8", model = "SM-X200")
        assertEquals(fixture("android-wire/pair-request.json"), encode("pairRequest", base, PairRequest.serializer()))
        assertEquals(fixture("android-wire/pair-request-code.json"), encode("pairRequest", base.copy(pairingCode = "012345"), PairRequest.serializer()))
    }

    @Test fun `real Android codec emits auth snapshot and audio bytes`() {
        assertEquals(fixture("android-wire/authenticate.json"), encode("authenticate", Authenticate(deviceId = "device-wire", clientNonce = "client-nonce-wire", timestampMs = 1_700_000_000_000, proof = "proof-wire"), Authenticate.serializer()))
        assertEquals(fixture("android-wire/request-snapshot.json"), encode("requestSnapshot", RequestSnapshot(), RequestSnapshot.serializer()))
        assertEquals(fixture("android-wire/audio-stream-start.json"), encode("audioStreamStart", AudioStreamStart(streamId = "stream-wire", sampleRate = 16000, sequenceStart = 0, captureTimestampMs = 1_700_000_000_000, frameDurationMs = 200), AudioStreamStart.serializer()))
    }

    @Test fun `real Android codec decodes Windows challenge bytes`() {
        val envelope = codec.decodeEnvelope(fixture("windows-wire/server-challenge.json"))
        assertEquals("challenge-wire", codec.decodePayload<ServerChallenge>(envelope).nonce)
    }
}

File: src/tests/androidWireCompatibility.test.ts
import { readFileSync } from 'node:fs';
import path from 'node:path';
import { describe, expect, it } from 'vitest';
import { ProtocolValidator } from '../../electron/server/ProtocolValidator';

const root = path.join(process.cwd(), 'shared', 'protocol', 'fixtures', 'android-wire');
const validator = new ProtocolValidator(path.join(process.cwd(), 'shared', 'protocol', 'schemas'));
const read = (name: string) => JSON.parse(readFileSync(path.join(root, name), 'utf8'));

describe('Android wire compatibility', () => {
  it.each([
    ['pair-request.json', 'pairing'],
    ['pair-request-code.json', 'pairing'],
    ['authenticate.json', 'pairing'],
    ['audio-stream-start.json', 'audio']
  ] as const)('accepts real Android codec fixture %s', (name, payloadKind) => {
    const envelope = read(name);
    validator.assert('envelope', envelope);
    validator.assert(payloadKind, envelope.payload);
  });

  it('preserves a leading zero in the six-digit code', () => {
    expect(read('pair-request-code.json').payload.pairingCode).toBe('012345');
  });

  it('accepts Android snapshot request routing envelope', () => {
    const envelope = read('request-snapshot.json');
    validator.assert('envelope', envelope);
    expect(envelope.payload).toEqual({ action: 'requestSnapshot' });
  });
});
